packages/backend/server/src/plugins/copilot/resolver.ts

```diff
@@ -724,6 +732,69 @@ export class CopilotResolver {
     }
   }
 
+  @Query(() => String, {
+    description:
+      'Apply updates to a doc using LLM and return the merged markdown.',
+  })
+  async applyDocUpdates(
+    @CurrentUser() user: CurrentUser,
+    @Args({ name: 'workspaceId', type: () => String })
+    workspaceId: string,
+    @Args({ name: 'docId', type: () => String })
+    docId: string,
+    @Args({ name: 'op', type: () => String })
+    op: string,
+    @Args({ name: 'updates', type: () => String })
+    updates: string
+  ): Promise<string> {
+    // Permission check: user must have Doc.Read access
+    const ac = this.ac as AccessController;
+    const docContent = await this.docReader.getDocMarkdown(
+      workspaceId,
+      docId,
+      true
+    );
+    if (!docContent || !docContent.markdown) {
+      throw new NotFoundException('Doc not found or empty');
+    }
+
+    const markdown = docContent.markdown.trim();
+
+    const canAccess = await ac
+      .user(user.id)
+      .workspace(workspaceId)
+      .doc(docId)
+      .can('Doc.Read');
+    if (!canAccess) {
+      throw new BadRequestException('No permission to read this doc');
+    }
+    // Get prompt for Apply Updates
+    const prompt = await this.promptService.get('Apply Updates');
+    if (!prompt) {
+      throw new BadRequestException('Prompt not found or invalid');
+    }
+
+    // Get LLM provider
+    const provider = await this.providerFactory.getProviderByModel(
+      prompt.model
+    );
+    if (!provider) {
+      throw new BadRequestException('No LLM provider available');
+    }
+
+    try {
+      return await provider.text({ modelId: prompt.model }, [
+        ...prompt.finish({}),
+        {
+          role: 'user',
+          content: `<code>${markdown}</code>\n<updates>${updates}</updates>`,
+        },
```

Comment:
can `prompt.finish({markdown, updates})` directly
then move the user message to prompt template and use {{markdown}} to render params as a message